planning/behavior_velocity_planner/autoware_behavior_velocity_blind_spot_module/src/time_to_collision.cpp

```diff
@@ -0,0 +1,239 @@
+    passage_time_intervals;
+  for (const auto & predicted_path : object.kinematics.predicted_paths) {
+    if (predicted_path.path.size() < 2) {
+      continue;
+    }
+    const double time_step = predicted_path.time_step.sec + predicted_path.time_step.nanosec * 1e-9;
+    const double horizon = time_step * static_cast<double>(predicted_path.path.size());
+    static constexpr double new_time_step = 0.1;
+    const auto precise_predicted_path = autoware::object_recognition_utils::resamplePredictedPath(
+      predicted_path, new_time_step, horizon);
+    const auto & shape = object.shape;
+
+    // search forward
+    std::optional<double> entry_time{};
+    for (const auto & [i, pose] : ranges::views::enumerate(precise_predicted_path.path)) {
+      const auto object_poly = autoware_utils_geometry::to_polygon2d(pose, shape);
+      if (boost::geometry::intersects(object_poly, lanelet::utils::to2D(line1).basicLineString())) {
+        entry_time = i * new_time_step;
+        break;
+      }
+    }
+
+    // search backward
+    std::optional<double> exit_time{};
+    for (const auto & [i, pose] :
+         ranges::views::enumerate(precise_predicted_path.path | ranges::views::reverse)) {
+      const auto object_poly = autoware_utils_geometry::to_polygon2d(pose, shape);
+      const double time = horizon - i * new_time_step;
+      if (entry_time && time < entry_time.value()) {
+        break;
+      }
+      if (boost::geometry::intersects(object_poly, lanelet::utils::to2D(line2).basicLineString())) {
+        exit_time = time;
+        break;
+      }
+    }
+    if (!exit_time) {
+      continue;
+    }
+    // in case the object is completely inside conflict_area, it is regarded entry_time = 0.0
+    passage_time_intervals.emplace_back(
+      entry_time.value_or(0.0), exit_time.value(), predicted_path);
+  }
+  return passage_time_intervals;
+}
+
+autoware_internal_planning_msgs::msg::SafetyFactor UnsafeObject::to_safety_factor() const
+{
+  autoware_internal_planning_msgs::msg::SafetyFactor factor;
+  factor.type = autoware_internal_planning_msgs::msg::SafetyFactor::OBJECT;
+  factor.object_id = object.object_id;
+  factor.predicted_path = predicted_path;
+  factor.ttc_begin = std::get<0>(object_passage_interval);
+  factor.ttc_end = std::get<1>(object_passage_interval);
+  factor.points.push_back(object.kinematics.initial_pose_with_covariance.pose.position);
+  factor.is_safe = false;
+  return factor;
+}
+
+std::optional<double> is_critical_for_ego(
+  const std::pair<double, double> & ego_passage_interval,
+  const std::pair<double, double> & object_passage_interval, const double ttc_start_margin,
+  const double ttc_end_margin)
+{
+  const auto & [ego_entry, ego_exit] = ego_passage_interval;
+  const auto & [object_entry, object_exit] = object_passage_interval;
+  // case1: ego will be still in conflict_area, when the object enters conflict_area
+  if (object_entry - ttc_start_margin < ego_exit) {
```

Comment:
This assumes `ego_entry < object_entry` (ego entered before the object) which can be wrong and cause false positives.

https://github.com/user-attachments/assets/e1044802-605a-44ca-bfa7-f5f37ca02ad1

Reply:
in f3d3f9802c8a7424d5775c7e7155f65633b8dc83, if `object_exit < ego_entry` this FP is ignored properly